src/graph/LdapAuthenticator.cpp

```diff
@@ -0,0 +1,290 @@
+
+    auto ret = ldap_simple_bind_s(ldap_, binddn.c_str(), bindPassword.c_str());
+    if (ret != LDAP_SUCCESS) {
+        ldap_unbind(ldap_);
+        return Status::Error("Perform initial LDAP bind for ldapbinddn \"%s\" on server "
+                             "\"%s\" failed.", binddn.c_str(), FLAGS_ldap_server.c_str());
+    }
+
+    // Build the filter
+    auto filter = buildFilter();
+
+    // Initiate an ldap search
+    ret = ldap_search_s(ldap_,
+                        FLAGS_ldap_basedn,
+                        0,  // The search scope, use LDAP_SCOPE_BASE
+                        filter,
+                        attributes_;
+                        0,
+                        &searchMessage);
+
+    if (ret != LDAP_SUCCESS) {
+        ldap_unbind(ldap_);
+        return Status::Error("Search LDAP for filter \"%s\" on server \"%s\" failed.",
+                             filtet.c_str(), FLAGS_ldap_server.c_str());
+    }
+
+    // Check the number of search result
+    auto retCount = ldap_count_entries(ldap_, searchMessage);
+    if (retCount ！= 1) {
+        ldap_unbind(ldap_);
+        ldap_msgfree(searchMessage);
+        if (retCount == 0) {
+            return Status::Error("LDAP user \"%s\" does not exist.", userName_.c_str());
+        } else {
+            return Status::Error("LDAP user \"%s\" is not unique.", userName_.c_str());
+        }
+    }
+
+    // Get distinguished name from search result
+    entry = ldap_first_entry(ldap_, searchMessage);
+    distName = ldap_get_dn(ldap_, entry);
+
+
+    if (!distName) {
+        ldap_unbind(ldap_);
+        ldap_msgfree(searchMessage);
+        return Status::Error("Get distinguished name for the first entry with filter \"%s\" "
+                             "on server \"%s\" failed.", filtet.c_str(),
+                             FLAGS_ldap_server.c_str());
+    }
+
+    dn = distName;
+    ldap_memfree(distName);
+    ldap_msgfree(searchMessage);
+
+    // Unbind and disconnect the first connection from the LDAP server
+    ret = ldap_unbind_s(ldap_);
+    if (ret != LDAP_SUCCESS) {
+        return Status::Error("Unbind failed after searching for user \"%s\" on server \"%s\".",
+                              userName_.c_str(), FLAGS_ldap_server.c_str());
+    }
+
+    /**
+     * Secondly, reinit LDAP connect to LADP server for the second connect.
+     * Using the distinguished name found on the first connection and the user's
+     * login password to LDAP authentication
+     */
+    auto reInit = initLDAPConnection();
+    if (!reInit.ok()) {
+        return reInit;
+    }
+
+    ret = ldap_simple_bind_s(ldap_, dn, password_);
+    if (ret != LDAP_SUCCESS) {
+        ldap_unbind(ldap_);
+        return Status::Error("LDAP login failed for user \"%s\" on server \"%s\"."
+                              dn.c_str(), FLAGS_ldap_server.c_str());
+    }
+
+    ldap_unbind(ldap_);
+    return true;
+}
+
+
+StatusOr<bool> LdapAuthenticator::directBindAuth() {
```

Comment:
Please make a research on how `Status`, `StatusOr<bool>`, `StatusOr<bool>.value()` are different.

Reply:
Yeah,
Status:      only has status.
StatusOr<bool>  has status or bool value， but cant have  both.
StatusOr<bool>.value():    when StatusOr<bool>.ok() is true， it has bool value and gets the bool value。